### PR TITLE
✨ Support IP Reuse for BMHs using Preallocations

### LINKED ISSUE #630: IP reuse feature support in CAPM3 using new M3IPClaim name format
**User Story**

As a [developer/user/operator] I would like to make IP addresses or IP allocations that will be attached to the nodes in the chain of (BMH <==> M3M <==> CAPI machine) Metal3 objects predictable. 

**Detailed Description**

As is now, an M3IPPool is an object representing a set of IP address pools to be used for IP address allocations. An M3IPClaim is an object representing a request for an IP address allocation. And, currently, the M3IPClaim name is structured based on the:

[M3DataName](https://github.com/metal3-io/cluster-api-provider-metal3/blob/main/docs/api.md#the-metal3data-object) + M3IPPoolName, i.e m3datatemplate-0-pool1. 

M3DataName is always created in a format of `M3DTemplateName-index`, so when m3m is created and M3Data is
requested, the index it will get is random, meaning during the upgrade,  `metal3machine0` and `metal3machine1` could get M3Datas created with indexes 0&1 respectively in their names, i.e: m3datatemplate-0&m3datatemplate-1 accordingly, and on the next upgrade indexes might be inverted as   `metal3machine0` could get `m3datatemplate-1` and  `metal3machine1` could get `m3datatemplate-0 
`

To reuse the same IP addresses from IPClaim allocated to metal3machine through M3Data during the upgrade, we can use a field called
`Preallocations map[claimName]Ipaddr` where `claimName` format is structured not the following the format of (`M3DataName+M3IPPoolName`) but following the format of  `BMH name+M3IPPoolName`, making it predictable so that
we can predict the claim Name we want to use for a specific BMH.

**Anything else you would like to add:**

 Not to break the current workflow of  M3IPClaim creation, we could introduce a feature gate that would be behind the controller, and based on that, user could opt using either the format of M3IPClaim name based on the BMH name or fallback using the current
format where M3IPClaim name is based on the M3Data name.

/kind feature

### PR DESCRIPTION
**Pull request by furkatgofurov7**
Fixes #630 
Related to: https://github.com/metal3-io/ip-address-manager/pull/119

### CODE REVIEW

**Comment by kashifest** [comment]
Can you please squash the BMH preallocation commits in one and run the integration test ? Have we tested this with the IPAM PR?

File: baremetal/utils.go
Comment (kashifest) [reply baremetal/utils.go:289]: Sorry ignore, I understand it wrongly, it is actually just wraping the error

**Comment by furkatgofurov7** [comment]
/hold
have to verify one use case locally

**Comment by furkatgofurov7** [comment]
> Can you please squash the BMH preallocation commits in one and run the integration test ? Have we tested this with the IPAM PR?

I will squash the commits and run the tests.

File: baremetal/metal3data_manager.go
Comment (lentzi90) [inline baremetal/metal3data_manager.go:591]: What is this doing, just assigning to itself?

File: baremetal/metal3data_manager.go
Comment (furkatgofurov7) [reply baremetal/metal3data_manager.go:590]: > This looks like public/exported variables with the capital "IP" to me. Maybe better to use `ipClaimsList` and `ipClaimWithLabels` to keep it similar to `ipClaim` and also show that it is private.

Yes, actually I have to change that to make it aligned. 



> Also, if err != nil we just continue, is this expected? A comment to explain would be good in that case. If I understand correctly this part with the labels is for preallocations, so it makes sense that if we don't have any we continue to check the "normal" case. But what if there was some temporary network issue? Should we not requeue then instead?

yes, we do try first to fetch and delete the claims with preallocation labels regardless or not we are using this feature (to prevent leaving up the resources in case the flag for enabling the feature was changed from ON to OFF in the bad timing let's say). As is we continue if we do not find any object to delete, but I do not know if we should requeue actually at least we do not do it in most of the places in the codebase.

File: baremetal/metal3data_manager.go
Comment (lentzi90) [reply baremetal/metal3data_manager.go:591]: Ah makes sense.

File: baremetal/metal3data_manager.go
Comment (lentzi90) [reply baremetal/metal3data_manager.go:590]: > yes, we do try first to fetch and delete the claims with preallocation labels regardless or not we are using this feature (to prevent leaving up the resources in case the flag for enabling the feature was changed from ON to OFF in the bad timing let's say).

:+1:

> As is we continue if we do not find any object to delete, but I do not know if we should requeue actually at least we do not do it in most of the places in the codebase.

How about checking if we get a "not found" error? In that case it is at least clear that we can continue without issue. If there is some other error I would suggest at least logging it, not sure if we need to requeue.

Comment (furkatgofurov7) [reply baremetal/metal3data_manager.go:590]: @lentzi90 deleteObject https://github.com/Nordix/cluster-api-provider-metal3/blob/9863d8f2805708088eb47866e877a78e75e00fe8/baremetal/utils.go#L112 already checks for "not found" error and returns nil if it meets that condition, otherwise returns error. What I did now is, I removed not found err check since that is part of the func already and return err when there is !not found error

Comment (furkatgofurov7) [reply baremetal/metal3data_manager.go]: > Yes, actually I have to change that to make it aligned.

Done

Comment (lentzi90) [reply baremetal/metal3data_manager.go]: Thanks, LGTM!

**Review by lentzi90 — APPROVED** [review]
This looks good to me now, just needs some squashing.

PS I don't have LGTM rights yet, but I think I may have enough commits now to get it :wink: :wink: :wink:

**Comment by kashifest** [comment]
> PS I don't have LGTM rights yet, but I think I may have enough commits now to get it 😉 😉 😉

@lentzi90 I have added your name here in the update Owners PR #710

**Comment by lentzi90** [comment]
/lgtm
:muscle:

**Comment by furkatgofurov7** [comment]
> have to verify one use case locally

/hold cancel
should be fine now. 

> Can you please squash the BMH preallocation commits in one and run the integration test ? Have we tested this with the IPAM PR?

done.

**Pull request merged**
